### ISSUE TITLE
[no-unnecessary-condition] false positive on RHS of LHS of nested conditional

### ISSUE THREAD

**Issue by astorije** [typescript-eslint/typescript-eslint#2128]
Apologies if this is a duplicate. I went through existing issues related to this rule and it didn't seem like this specific case was already reported.

**Repro**

[Playground](https://typescript-eslint.io/play#ts=5.3.2&fileType=.tsx&code=GYVwdgxgLglg9mABMAhjANgZwBTDnAfgC5EAjfdAUxTABoyUAnYsi6sASkQG8AoRRI0pQQjJLnyIAZFMQByPHDlcAPisTZSTabLlbGygNy8AvryA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

```JSON
{
  "rules": {
    "@typescript-eslint/no-unnecessary-condition": "error"
  }
}
```

<details>
<summary>tsconfig.json</summary>

```json5
{
  "compilerOptions": {
    /* Basic Options */
    "target": "es6", // Specify ECMAScript target version.
    "lib": ["dom", "es2015", "es2017", "esnext.asynciterable"],
    "jsx": "react", // Specify JSX code generation.
    "declaration": true, // Generates corresponding '.d.ts' file.
    "composite": true, // Ensure TypeScript can determine where to find the outputs of the referenced project to compile project.
    "preserveWatchOutput": true, // Keep outdated console output in watch mode instead of clearing the screen.

    /* Strict Type-Checking Options */
    "strict": true, // Enable all strict type-checking options.

    /* Additional Checks */
    "noUnusedLocals": true, // Report errors on unused locals.
    "noUnusedParameters": true, // Report errors on unused parameters.
    "noImplicitReturns": true, // Report error when not all code paths in function return a value.
    "noFallthroughCasesInSwitch": true, // Report errors for fallthrough cases in switch statement.

    /* Module Resolution Options */
    "moduleResolution": "node", // Specify module resolution strategy.
    "allowSyntheticDefaultImports": true, // Allow default imports from modules with no default export. This does not affect code emit, just typechecking.
    "esModuleInterop": true, // Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports.
    "forceConsistentCasingInFileNames": true // Disallow inconsistently-cased references to the same file.
  }
}
```

</details>

```ts
function fails(foo?: boolean, bar?: boolean) {
  return (foo && 'foo') || (bar && 'bar');
}
```

**Expected Result**

I do not expect this snippet to fail this rule, as this was the case in v2.34.

**Actual Result**

This is flagged as a violation:

```ts
error: Unnecessary conditional, value is always truthy (@typescript-eslint/no-unnecessary-condition)
      | function example(foo?: boolean, bar?: boolean) {
   >  |   return (foo && 'foo') || (bar && 'bar');
      |                  ^
      | }
```

**Additional Info**

This passes fine if turned into a set of ternaries:

```ts
function passes(foo?: boolean, bar?: boolean) {
  return foo ? 'foo' : bar ? 'bar' : undefined;
}
```

(I realize those are not _exactly_ identical because of `false` being in the return type of the first example, but this is just for the sake of demonstration 😅)

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.0.2` |
| `@typescript-eslint/parser`        | `3.0.2` |
| `TypeScript`                       | `3.9.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.14.0` |
| `yarn`                              | `1.22.4` |

**Comment by bradzacher** [comment]
this is likely a side-effect of removing the `ingoreRHS` option.

this is a hard problem to solve because it involves building complete understanding of short-circuiting into the rule.

---

As an engineer - IMO this is not good code, because it's really hard to understand. Whilst it's short and concise, it is "too clever", and sacrifices understandability.

IMO your ternary is much clearer, or even clearer still would be two `if`s:
```ts
function passes(foo?: boolean, bar?: boolean) {
  if (foo) {
    return 'foo';
  }
  if (bar) {
    return 'bar';
  }
  return false;
}
```
Code is read many more times than it's written, so optimising for clarity and readability is paramount IMO.

**Comment by ehahn9** [comment]
(newbie here) - the removal of `ignoreRHS` has a fairly disastrous impact on code like this (which IMO is super common in front-end markup):

```html
<div>
  {error && <Error message={error.message} />}
</div>
```
In this (common?) idiom, we're using `&&` as a conditional, since `false` is okay to present to the markup engine (React in my case).

The only work-arounds are both unappealing: 1) disable the rule altogether (sad face) or 2) litter exceptions all over markup (also sad face). `ignoreRHS` was really a wonderful thing.

Regression?

**Comment by bradzacher** [comment]
@ehahn9 - `ignoreRHS` was removed intentionally because it was inherently flawed. You do not want to always ignore the RHS of a logical expression.  
If you always ignore it, then you'll get false negatives like: `if (foo && true) {}`

Instead you want to be smart about it - you want to ignore the RHS when it doesn't matter, i.e. when it's _not_ used in a boolean context. If you add this condition to the option, then there's actually no reason to ever have the option turned off, so then there's no need to have the option.

---

Additionally, I'm not able to reproduce any error for the given code. The logical expression is not in a boolean context, so the rule should (and does, from my testing) ignore the RHS.

I tested the following code and it passed fine:
```ts
import React from 'react';
declare const error: Error | null;
const x = <div>{error && <Error message={error.message} />}</div>;
```

If I were to hazard a guess, I'd say that you don't have `strictNullChecks` turned on. This is a common cause of problems for users of the rule.

**Comment by ehahn9** [comment]
@bradzacher thanks for the quick reply - much appreciated.  `strictNullChecks` definitely turned on so I'll do more to narrow down (and make a minimal repo to pursue once I'm more certain).  Feel free to close this issue in the meantime. Apologies if this turns out to be a red herring.

**Comment by astorije** [comment]
@bradzacher, thanks for the quick replies.

I agree that there are more than one way to do what I described in the original post, and I fully agree that readability and clarity trump one-liner cleverness by a long shot, however those were just out-of-context examples to keep things simple and should not be used to determine the quality of the underlying code 😅

This is actually used in a React prop, so while i agree that `if`s statement make for the clearest, they cannot be used here without extra hoops. A slightly less contrived example (but still a lot simplified) would look like this:

```tsx
interface FoobarProps {
  foo?: boolean;
  bar?: boolean;
}

function Foobar({ foo, bar }: FoobarProps) {
  return <Component value={(foo && 'Foo') || (bar && 'Bar')} />;
}
```

It's not actually particularly clever, and it's rather idiomatic and fairly common: when the first flag is set, pass the first value, or when the second flag is set, pass the second value.
The ternary equivalent would still look like:

```tsx
  return <Component value={(foo ? 'Foo' : bar ? 'Bar' : undefined)} />;
```

This might be subjective, but I don't find this any more readable, actually the opposite.
And the `if` version would require moving the value to its own function, which is an unnecessary indirection IMO.

@ehahn9, this is indeed a very common way to conditionally render blocks in React. We in fact use this a lot and, as far as I can tell, were not impacted by this rule when migrating to v3 (with `strictNullChecks` on), so I'm curious to hear what you discover.

---

@bradzacher, stepping back a little, it seems the rule is still incorrect here, regardless of how this can be improved. `foo` can still be `false` or `undefined`, so definitely not always truthy.
That said, I do appreciate the feedback on clarity and always welcome new takes that lead to better code 😊 Thanks!

**Comment by Retsam** [comment]
Yeah, I'm a React developer, so I definitely had `{cond && <Component>}` in mind when removing the `ignoreRHS` rule.  The new logic is to only check the RHS when the expression is used in a conditional context, (inside an if statement, as part of a bigger boolean expression). 

And I can confirm we didn't get false-positives on this from our React codebase.

---

I do think the root issue here is hard to fix.  It's valid in this case because `"foo"` isn't really being used for its truthiness, but for its value, and it's going to be hard to distinguish those cases.  

This is reasonable code that could hypothetically be permitted by this linter rule:

```ts
console.log((a && "a") || b)
```

But these two examples are certainly mistakes:

```ts
// Since this is only being used for its truthiness, `&& "a"` changes the value, but doesn't change the truthiness
if(a && "a" || b);
// This case, the "a" value is just going to get overridden by `b`, so it's unnecessary as well.
console.log(a && "a" && b) 
```

I don't believe there's any reasonable way to permit the first while flagging the second examples - you'd have to determine when the conditional is being used for its truthyness or for its value.

And to avoid false-negatives like the third example, it'd have to be very precise about `&&` and `||`.  (An always truthy value is reasonable as the left-hand side of `||`, but not as the left-hand side of `&&`)

---

I think this is likely just going to have to be a known exception: if you're adding conditions just to change the value and not to change the truthiness, you might need to flag the line as an exception.  

I'd prefer that over reintroducing a flag like `ignoreRHS` which introduces false-negatives to reduce a few false-positives.  This case feels pretty niche to me, so I don't think it's worth the cost of adding an option.

---

To the specific instance, I do think the ternary is cleaner - a ternary's "if/else" is just a simpler concept than the short-circuiting based version.  

This may be a suggestion based on an simplified example, but if possible, I'd suggest writing this component as:

```ts
interface FoobarProps {
  foobar?: "Foo" | "Bar" | undefined
}

function Foobar({ foobar }: FoobarProps) {
  return <Component value={foobar} />;
}
```

Not only does it avoid this whole question of ternary vs. conditional, but I think it clarifies the contract as the `foo` and `bar` options really aught to be mutually exclusive: `<Foobar foo bar /}` really doesn't make sense as `foo` will override `bar`.  (Again, this might only apply to the simplified example)

**Comment by ehahn9** [comment]
Thanks all for the great exploration. I'm 100% on board, so no need to keep this open for me. BTW, lest I seem at all ungrateful, @typescript-eslint has been an absolute delight to use.

**Comment by astorije** [comment]
This is all great feedback, thank you so much! @Retsam, it makes sense why the rule would flag this, why it's a complex problem to fix, and why the `ignoreRHS` was more of a curse than a blessing.

In the meantime, ternary it is! As I upgrade across our projects, I'll see how this pattern is wide-spread or not, hopefully not.

> you might need to flag the line as an exception

Unrelated to this issue or project, but we try to discourage line overrides when there is another way, as we've noticed doing so builds muscle memory for people to just disable rules when they see violations. It is a dev culture and we're trying to fix it, but old habits die hard!
Hence the acceptable ternary here, even if I personally don't find it as readable since it's not about personal opinions.

> it clarifies the contract as the `foo` and `bar` options really aught to be mutually exclusive [...] (Again, this might only apply to the simplified example)

Ahah, it is in fact simplified: in the context this is taken from, `foo` and `bar` aren't even mutually exclusive (but there is a precedence between them), but more importantly, this is coming from an API outside our boundaries.
Besides, I was trying to illustrate the issue around those operators using a minimal reproductive example, and not about this foobar code 😅

> BTW, lest I seem at all ungrateful, @typescript-eslint has been an absolute delight to use.

I sure hope I don't come across ungrateful either: I'm a huge fan of ESLint, am really happy of being able to convert our projects from TSLint to ESLint among my organization, and therefore incredibly grateful for all your hard work and amazing feedback. Thanks, all!
